tests/unitary_calculator_testfixture.h

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef UNITARY_CALCULTOR_TESTFIXTURE_H_
+#define UNITARY_CALCULTOR_TESTFIXTURE_H_
+
+#include <complex>
+
+#include "gtest/gtest.h"
+
+namespace qsim {
+
+namespace unitary {
+
+namespace {
+
+template <typename UnitarySpace, typename Unitary>
+void FillMatrix(UnitarySpace& us, Unitary& u, int n) {
+  for(int i =0; i < (1 << n); i++){
+    for(int j =0;j < (1 << n); j++) {
+      us.SetEntry(u, i, j, 2 * i * (1 << n) + 2 * j,
+        2 * i * (1 << n) + 2 * j + 1);
+    }
+  }
+}
+
+}  // namespace
+
+constexpr char provider[] = "unitary_calculator_test";
+
+template <typename UC>
+void TestApplyGate1() {
+  const int n_qubits = 3;
+  UC uc(n_qubits, 1);
+  using UnitarySpace = typename UC::UnitarySpace;
+  using Unitary = typename UC::Unitary;
+
+  UnitarySpace us(n_qubits, 1);
+  Unitary u = us.CreateUnitary();
+
+  float ref_gate[] = {1,2,3,4,5,6,7,8};
+
+  // Test applying on qubit 0.
+  FillMatrix(us, u, n_qubits);
+  // clang-format off
+  float expected_mat_0[] = {
```

Comment:
It's not immediately clear why this hardcoded matrix is the expected output. Either generating it in a self-explanatory way or adding a comment explaining it would be helpful.

Same for the variants below.

Reply:
I find the code to be pretty self documenting in this case. We are doing the C++ equivalent of:

```python
a = gate_matrix
b = unitary_matrix
res = np.kron(np.eye(4), a) @ b
```
How do you think we should make it better ?

Reply:
I agree that the operations are clear. My concern was with `expected_mat_0`, but I suppose it can be derived from the structure of the test, so it's OK to leave as-is.